### ISSUE TITLE
Deadlock: Multi-VU many iterations

### ISSUE THREAD

**Issue by ankur22** [grafana/xk6-browser#971]
While running the following test script:

<details>
<summary>Test script</summary>

```js
import { check } from 'k6';
import { browser } from 'k6/experimental/browser';

export const options = {
  scenarios: {
    ui: {
      executor: 'shared-iterations',
      vus: 5,
      iterations: 1000,
      maxDuration: '40m',
      options: {
        browser: {
          type: 'chromium',
        },
      },
    },
  },
  thresholds: {
    checks: ["rate==1.0"]
  }
}

export default async function () {
  const context = browser.newContext();
  const page = context.newPage();

  try {
    // Goto front page, find login link and click it
    await page.goto('https://test.k6.io/', { waitUntil: 'networkidle' });
    await Promise.all([
      page.waitForNavigation(),
      page.locator('a[href="/my_messages.php"]').click(),
    ]);
    // Enter login credentials and login
    page.locator('input[name="login"]').type('admin');
    page.locator('input[name="password"]').type('123');
    // We expect the form submission to trigger a navigation, so to prevent a
    // race condition, setup a waiter concurrently while waiting for the click
    // to resolve.
    await Promise.all([
      page.waitForNavigation(),
      page.locator('input[type="submit"]').click(),
    ]);
    check(page, {
      'header': page.locator('h2').textContent() == 'Welcome, admin!',
    });
  } finally {
    page.close();
  }
}
```

</details>

The test has to be aborted since it goes over the 40m limit. Investigate why this happens and resolve it.

![Image](https://github.com/grafana/xk6-browser/assets/1112428/a002bb66-f3ce-4bf4-b7eb-557500f5337a)

This is the screenshot of the dashboard of the test run. It would seem that all but one of the VUs completed all the allotted iterations. One vu was stuck in a deadlock, which meant the test run was waiting for it to complete the iteration, which it never did, that eventually resulted in an abort.

---

```[tasklist]
### Tasks
- [x] Collect every instance of this issue occurring
- [x] Investigate the root cause
- [x] Fix
- [x] Test
- [x] Polish
- [x] Retest
- [ ] https://github.com/grafana/xk6-browser/pull/1219
- [ ] Release
- [ ] Update the release notes
```

**Comment by ankur22** [comment]
## Investigation

### Rerun test with different params

The test was ran again, with the fields amended to the values shown in the table:

| Iterations | Max duration | Success (not aborted) |
| - | - | - |
| 5 | 5m | true |
| 10 | 10m | true |
| 100 | 30m | true |
| 1000 | 50m | true |

I was unable to replicate the issue. All tests ended within the allotted max duration.

**Comment by ankur22** [comment]
A temporary fix can be found in https://github.com/grafana/xk6-browser/pull/979.

**Comment by inancgumus** [comment]
Could there be a race in handling iteration events? 🤔 

From the stack trace, it seems like both `page.Close` and closing of the browser happens concurrently.

```
goroutine 56 [select]:
github.com/grafana/xk6-browser/common.(*Browser).Close(0xc00067bc80)
	github.com/grafana/xk6-browser@v1.0.0-rc0.0.20230711104523-a8ebe8cc24f6/common/browser.go:464 +0x313
github.com/grafana/xk6-browser/browser.(*browserRegistry).deleteBrowser(0xc001cb2c80, 0xc001cb2840?)
	github.com/grafana/xk6-browser@v1.0.0-rc0.0.20230711104523-a8ebe8cc24f6/browser/registry.go:310 +0x9c
github.com/grafana/xk6-browser/browser.(*browserRegistry).handleIterEvents(0xc001cb2c80, 0x1a0fea0?, 0xc001480d80)
	github.com/grafana/xk6-browser@v1.0.0-rc0.0.20230711104523-a8ebe8cc24f6/browser/registry.go:269 +0x1c5
created by github.com/grafana/xk6-browser/browser.newBrowserRegistry
	github.com/grafana/xk6-browser@v1.0.0-rc0.0.20230711104523-a8ebe8cc24f6/browser/registry.go:219 +0x311
```

cc: @ankur22 @ka3de

**Comment by inancgumus** [comment]
Right. But, in the meantime, might we be closing the browser and this might be racing with `page.Close`? It's weird that it occurs only on `pagehide`. It should happen elsewhere too. Why only there? :)

**Comment by ankur22** [comment]
> Right. But, in the meantime, might we be closing the browser and this might be racing with `page.Close`? It's weird that it occurs only on `pagehide`. It should happen elsewhere too. Why only there? :)

I don't have the answer for why it's occurring in this scenario, I was hoping that it was due to the overloaded chrome instance, however other APIs (such as page.goto) they eventually receive a response when the default timeout is set very high, which would mean that chrome does indeed respond to CDP requests when it is under heavy load, and doesn't drop messages.

When the evaluate on `pagehide` event blocks waiting on the response from chrome, that VU is stuck indefinitely, and it's stuck on the "main thread" of that VU, putting a pause on the whole underlying JS event loop. At this point even if k6 were to signal an `iterEnd`, the stuck VU wouldn't be able to handle it. So I can't see how there could be a race condition between the browser closing and page closing 🤔

**Comment by inancgumus** [comment]
It seems we need to look deeper into this issue since you have very good points and I don't have the best answer at this moment :(

**Comment by ankur22** [comment]
After working with this fix against a remote chrome instance, we've now been able to reproduce the issue with `page.goto` as well as with the `evaluate` on `page.close`. We did this by setting the `page`'s `defaultTimeout` to a very high value (30mins), which confirmed the VU deadlocks waiting for a response from chrome. When the timeouts are left at 30secs, the timeouts occur as expected and we do not see deadlocks. As far as I can tell this still leaves us with the these possibilities for the root cause of the problem:

1. Chrome isn't returning a response for some of these CDP requests.
2. Chrome is returning a malformed request which we're not propagating to all listening handlers. I've looked at the code and it seems to be set up correctly [here](https://github.com/grafana/xk6-browser/blob/4485c6d0caf9276b3a772e708a23f526a23f665c/common/connection.go#L371-L372) to catch malformed responses and seems to forward all other `msg`s correctly to all listening handlers.
3. We have a deadlock somewhere in xk6-browser.

One more point to make is that the remote chrome instances were **not** under a lot of load.
